File: src/studica_control/src/components/examples/python/ultrasonic_example.py
#!/usr/bin/env python3
"""Ultrasonic — subscribe to range readings.

Run:  python3 ultrasonic_example.py
Requires: studica_launch.py running, ultrasonic enabled in params.yaml

Topic:   subscribes to 'ultrasonic_range' (sensor_msgs/Range) — set ultrasonic.<sensor>.topic: ultrasonic_range in params.yaml
Service: '/<sensor_name>/ultrasonic_cmd' (SetData)
  Commands: get_distance, get_distance_inches, get_distance_millimeters
"""
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Range


class UltrasonicExample(Node):
    def __init__(self):
        super().__init__('ultrasonic_example')
        self.sub = self.create_subscription(Range, 'ultrasonic_range', self.on_range, 10)
        self.get_logger().info('Listening on ultrasonic_range topic...')

    def on_range(self, msg):
        self.get_logger().info(f'Range: {msg.range:.3f} m  (min: {msg.min_range}, max: {msg.max_range})')


def main():
    rclpy.init()
    rclpy.spin(UltrasonicExample())
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/studica_control/src/components/examples/cpp/ultrasonic_example.cpp
/*
 * ultrasonic_example.cpp
 *
 * Subscribe to HC-SR04 range readings published by the ultrasonic component.
 * Prints distance in metres; warns when the reading is out of sensor range.
 *
 * Run:      ros2 run studica_control ultrasonic_example
 * Requires: studica_launch.py running, ultrasonic enabled in params.yaml
 *           sensors: ["ultrasonic"]
 *
 * Topic subscribed: /ultrasonic/range  (sensor_msgs/Range)
 * Service:          /ultrasonic/ultrasonic_cmd  (studica_control/SetData)
 *   Commands: get_distance, get_distance_inches, get_distance_millimeters
 */

#include <cmath>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/range.hpp"

static const std::string SENSOR = "ultrasonic";

class UltrasonicExample : public rclcpp::Node {
public:
    UltrasonicExample() : Node("ultrasonic_example") {
        sub_ = create_subscription<sensor_msgs::msg::Range>(
            "/" + SENSOR + "/range", 10,
            std::bind(&UltrasonicExample::on_range, this, std::placeholders::_1));
        RCLCPP_INFO(get_logger(), "listening on /%s/range", SENSOR.c_str());
    }

private:
    void on_range(const sensor_msgs::msg::Range::SharedPtr msg) {
        if (std::isinf(msg->range)) {
            RCLCPP_WARN(get_logger(), "out of range (valid: %.2f – %.2f m)",
                        msg->min_range, msg->max_range);
        } else {
            RCLCPP_INFO(get_logger(), "range: %.3f m", msg->range);
        }
    }

    rclcpp::Subscription<sensor_msgs::msg::Range>::SharedPtr sub_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<UltrasonicExample>());
    rclcpp::shutdown();
    return 0;
}
